### PR TITLE
Closed #2583: `client_dtypes_test` conversion for new framework

### LINKED ISSUE #2583: `client_dtypes_test.py` conversion for new test framework
Convert `client_dtypes_test.py` to work in the new test framework. The existing file should be maintained since we will be doing a complete swap once all tests are converted. Only the existing tests need to be converted, there should not be any additional testing that needs to be added.

### PR DESCRIPTION
**Pull request by jaketrookman**
This PR (Closes #2583) updates `client_dtypes_test` to the new testing framework.

### CODE REVIEW

File: PROTO_tests/tests/client_dtypes_test.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        ans = ak.is_ipv4(ak.array(x, ak.uint64))
+        assert ans.to_list() == [i >= 5 for i in range(10)]
+
+        with pytest.raises(TypeError):
+            ak.is_ipv4(ak.array(x, ak.float64))
+
+        with pytest.raises(RuntimeError):
+            ak.is_ipv4(ak.array(x, dtype=ak.uint64), ak.arange(2, dtype=ak.uint64))
+
+    def test_is_ipv6(self):
```
Comment (TheOddCzar789) [inline PROTO_tests/tests/client_dtypes_test.py:144]: Do we want to use pytest.prob_size here?

File: PROTO_tests/tests/client_dtypes_test.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+class TestClientDTypeTests:
+    """
+    Note: BitVector operations are not tested here because the class is
+    only a wrapper on a pdarray to display as such.
+    The class does not actually store values as bit-values,
+    it only converts to a bit representation for display.
+    Thus, pdarray testing covers these operations.
+    """
+
+    def test_BitVector_creation(self):
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:18]: could we parameterize this on `[ak.int64, ak.uint64]` to avoid some of the dupication?

File: PROTO_tests/tests/client_dtypes_test.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names, separator="abc")
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names)
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names, pad="|!~", separator="//")
+
+    def test_ipv4_creation(self):
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:91]: again could we parameterize this on `[ak.int64, ak.uint64]`?

File: PROTO_tests/tests/client_dtypes_test.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        # Test handling of int64 input
+        ip_list = ak.array([3232235777], dtype=ak.int64)
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:93]: would it be possible to add more values to this? I believe there are also IP modules in Python that we could used for validation as opposed to hard-coding. If we use one of these, we may need to add it to the developer yaml.

File: PROTO_tests/tests/client_dtypes_test.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            ipv4 = ak.IPv4(ak.array([3232235777.177]))
+
+        # Test handling of python dotted-quad input
+        ipv4 = ak.ip_address(["192.168.1.1"])
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == ["192.168.1.1"]
+        assert ipv4.dtype == ak.bitType
+
+    def test_ipv4_normalization(self):
+        ip_list = ak.array([3232235777])
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:120]: again could we expand this to test against more than 1 IP?

**Review by Ethan-DeBandi99 — COMMENTED** [review]
I think there are a few spots where adding parameterization would minimize the amount of code. Also, it would be good to perform validation against more than 1 IP address and possibly use an existing framework for validation.

File: PROTO_tests/tests/client_dtypes_test.py (hunk 4 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated, the `ipaddress` library come already installed

File: PROTO_tests/tests/client_dtypes_test.py (hunk 5 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 3 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 2 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        assert isinstance(bv, client_dtypes.BitVector)
+        assert bv.to_list() == bv_answer
+        assert bv.dtype == ak.bitType
+
+        # fail on argument types
+        with pytest.raises(TypeError):
+            bv = ak.BitVector(17, width=4)
+        arr = ak.array([1.1, 8.3])
+        with pytest.raises(TypeError):
+            bv - ak.BitVector(arr)
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:53]: I would move these into a `test_bit_vector_error_handling` test. As currently configured, these will run identically twice.

File: PROTO_tests/tests/client_dtypes_test.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+            f = ak.Fields(values, names, separator="abc")
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names)
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names, pad="|!~", separator="//")
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    @pytest.mark.parametrize("ip", IPS)
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:100]: I don't think we need to parameterize this, we can test them all at once as one array. It will just require list comprehension for the validation using the `ipaddress` module.

File: PROTO_tests/tests/client_dtypes_test.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        with pytest.raises(TypeError):
+            ipv4 = ak.IPv4(ak.array([ip + 0.177]))
+
+        # Test handling of python dotted-quad input
+        ipv4 = ak.ip_address([format(ipaddress.IPv4Address(ip))])
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == [format(ipaddress.IPv4Address(ip))]
+        assert ipv4.dtype == ak.bitType
+
+    @pytest.mark.parametrize("ip", IPS)
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:120]: Again, I don't think these should be parameterized.

**Review by Ethan-DeBandi99 — CHANGES_REQUESTED** [review]
Few other small things.

File: PROTO_tests/tests/client_dtypes_test.py (hunk 4 above)
Comment (Ethan-DeBandi99) [reply PROTO_tests/tests/client_dtypes_test.py]: Additional comments on this, but they are below. Resolving.

File: PROTO_tests/tests/client_dtypes_test.py (hunk 5 above)
Comment (Ethan-DeBandi99) [reply PROTO_tests/tests/client_dtypes_test.py]: Additional comments on this, but they are below. Resolving.

File: PROTO_tests/tests/client_dtypes_test.py (hunk 8 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 7 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 6 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

**Review by Ethan-DeBandi99 — APPROVED** [review]

**Review by TheOddCzar789 — APPROVED** [review]

File: PROTO_tests/tests/client_dtypes_test.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    """
+    Note: BitVector operations are not tested here because the class is
+    only a wrapper on a pdarray to display as such.
+    The class does not actually store values as bit-values,
+    it only converts to a bit representation for display.
+    Thus, pdarray testing covers these operations.
+    """
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_BitVector_creation(self, int_types):
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:23]: we should be consistent in our test names
```suggestion
    def test_bit_vector_creation(self, int_types):
```

File: PROTO_tests/tests/client_dtypes_test.py
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:1]: I would like to have a test added that does some edge case testing. i.e. calling `ak.BitVector` on things in the `2**63..2**64-1` range

File: PROTO_tests/tests/client_dtypes_test.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names, pad="|!~", separator="//")
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_ipv4_creation(self, int_types):
+        ip_list = ak.array(IPS, dtype=int_types)
+        ipv4 = ak.IPv4(ip_list)
+
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == [format(ipaddress.IPv4Address(ip)) for ip in IPS]
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:106]: I don't love the use of `format` here (and in the other spots that use it). I feel like it's not super clear what the builtin is doing. I think you can achieve the same result using `.compressed`
```suggestion
        assert ipv4.to_list() == [ipaddress.IPv4Address(ip).compressed for ip in IPS]
```

File: PROTO_tests/tests/client_dtypes_test.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        with pytest.raises(TypeError):
+            ipv4 = ak.IPv4(f"{IPS}")
+        with pytest.raises(TypeError):
+            ipv4 = ak.IPv4(ak.array([IPS + 0.177]))
+
+        # Test handling of python dotted-quad input
+        ipv4 = ak.ip_address([format(ipaddress.IPv4Address(ip)) for ip in IPS])
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == [format(ipaddress.IPv4Address(ip)) for ip in IPS]
+        assert ipv4.dtype == ak.bitType
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:118]: we seem to use the same list comprehension over and over. We should save it off and reuse it
```suggestion
        py_ips = [ipaddress.IPv4Address(ip).compressed for ip in IPS]
        assert isinstance(ipv4, ak.IPv4)
        assert ipv4.to_list() == py_ips
        assert ipv4.dtype == ak.bitType

        with pytest.raises(TypeError):
            ipv4 = ak.IPv4(f"{IPS}")
        with pytest.raises(TypeError):
            ipv4 = ak.IPv4(ak.array([IPS + 0.177]))

        # Test handling of python dotted-quad input
        ipv4 = ak.ip_address(py_ips)
        assert isinstance(ipv4, ak.IPv4)
        assert ipv4.to_list() == py_ips
        assert ipv4.dtype == ak.bitType
```

File: PROTO_tests/tests/client_dtypes_test.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+        with pytest.raises(TypeError):
+            ak.is_ipv4(ak.array(x, ak.float64))
+
+        with pytest.raises(RuntimeError):
+            ak.is_ipv4(ak.array(x, dtype=ak.uint64), ak.arange(2, dtype=ak.uint64))
+
+    @pytest.mark.parametrize("size", pytest.prob_size)
+    def test_is_ipv6(self, size):
+        x = [random.getrandbits(128) for i in range(size)]
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:149]: since `i` is never used, python devs would normally name it `_` to indicate that, so
```suggestion
        x = [random.getrandbits(128) for _ in range(size)]
```

File: PROTO_tests/tests/client_dtypes_test.py (hunk 9 above)
Comment (pierce314159) [reply PROTO_tests/tests/client_dtypes_test.py:23]: same goes for `test_Field_creation` below

Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py:1]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 10 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 11 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 12 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 13]
```diff
@@ -52,8 +52,14 @@ def test_bit_vector_error_handling(self):
         arr = ak.array([1.1, 8.3])
         with pytest.raises(TypeError):
             ak.BitVector(arr)
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_bit_vector_edge_case(self, int_types):
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py]: we need to rerun black on this file because there should be a new line here

File: PROTO_tests/tests/client_dtypes_test.py [hunk 14]
```diff
@@ -52,8 +52,14 @@ def test_bit_vector_error_handling(self):
...
+        arr = ak.arange(2**64-10, 2**64-1, dtype=int_types)
+        bv = ak.BitVector(arr, width=3)
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:58]: we need the `width` to be much larger here to actually capture the full `bitVector` for numbers this large

also doesn't look like we're actually comparing these values to anything, we're just checking that they are `BitVector`s

**Review by pierce314159 — CHANGES_REQUESTED** [review]

File: PROTO_tests/tests/client_dtypes_test.py (hunk 14 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 13 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            f = ak.Fields(values, names, separator="abc")
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names)
+
+        with pytest.raises(ValueError):
+            f = ak.Fields(values, names, pad="|!~", separator="//")
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_ipv4_creation(self, int_types):
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:115]: It would be beneficial to have a "general" version of this where we also use `pytest.prob_size` and create IPs over some range, or even randomly generated integers. While this test is good for testing known examples, we need to expand the coverage a bit. I would recommend either adding an additional test or replacing this one in order to get better coverage.

File: PROTO_tests/tests/client_dtypes_test.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+            ipv4 = ak.IPv4(ak.array([IPS + 0.177]))
+
+        # Test handling of python dotted-quad input
+        ipv4 = ak.ip_address(py_ips)
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == py_ips
+        assert ipv4.dtype == ak.bitType
+
+    def test_ipv4_normalization(self):
+        ip_list = ak.array(IPS)
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/client_dtypes_test.py:136]: Again, I would really like this to be run on a larger range than just 4 numbers. Could we update this similar to my previous comment to use randint or something like that?

**Review by Ethan-DeBandi99 — CHANGES_REQUESTED** [review]
Few things to update

File: PROTO_tests/tests/client_dtypes_test.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+import pytest
+
+import arkouda as ak
+from arkouda import client_dtypes
+
+INT_TYPES = [ak.int64, ak.uint64]
+IPS = [3232235777, 2222222222, 1234567890, 7]
+
+
+class TestClientDTypeTests:
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:13]: This name says `Test` twice and needs to be updated
```suggestion
class TestClientDTypes:
```

File: PROTO_tests/tests/client_dtypes_test.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+        py_ips = [ipaddress.IPv4Address(ip).compressed for ip in IPS]
+
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == py_ips
+        assert ipv4.dtype == ak.bitType
+
+        with pytest.raises(TypeError):
+            ipv4 = ak.IPv4(f"{IPS}")
+        with pytest.raises(TypeError):
+            ipv4 = ak.IPv4(ak.array([IPS + 0.177]))
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:127]: these aren't testing the same things as the original. The second one in particular is testing something very different and just happens to give the same error as the following shows
```python
# the original does something analogous to this, which creates 
# a float pdarray the type error is just because only except int pdarrays
>>> ak.array([IPS[0] + 0.177])
array([3232235777.177])

# the suggest change raise a type error because it cannot create a pdarray
>>> ak.array([IPS + 0.177])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
----> 1 ak.array([IPS + 0.177])

TypeError: can only concatenate list (not "float") to list
```

both of these should be updated. I think the following should be enough to cover the original tests
```suggestion
        with pytest.raises(TypeError):
            ipv4 = ak.IPv4(f"{IPS[0]}")
        with pytest.raises(TypeError):
            ipv4 = ak.IPv4(ak.array([IPS[0] + 0.177]))
```

**Review by pierce314159 — CHANGES_REQUESTED** [review]

File: PROTO_tests/tests/client_dtypes_test.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
+import ipaddress
+import random
+
+import pytest
+
+import arkouda as ak
+from arkouda import client_dtypes
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    """
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_bit_vector_creation(self, int_types):
+        bv_answer = ["...", "..|", ".|.", ".||"]
+        bv_rev_answer = ["...", "|..", ".|.", "||."]
+
+        arr = ak.arange(4, dtype=int_types)
+        bv = ak.BitVector(arr, width=3)
+        assert isinstance(bv, client_dtypes.BitVector)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+        assert bv.to_list() == bv_answer
+        assert bv.dtype == ak.bitType
+
+        # Test reversed
+        arr = ak.arange(4, dtype=int_types)
+        bv = ak.BitVector(arr, width=3, reverse=True)
+        assert isinstance(bv, client_dtypes.BitVector)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+        assert bv.to_list() == bv_rev_answer
+        assert bv.dtype == ak.bitType
+
+        # test use of vectorizer function
+        arr = ak.arange(4, dtype=int_types)
+        bvectorizer = ak.BitVectorizer(3)
+        bv = bvectorizer(arr)
+        assert isinstance(bv, client_dtypes.BitVector)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    def test_bit_vector_edge_case(self, int_types):
+        bv_answer = [
+            "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||..",
+            "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|",
+            "|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.",
+        ]
+
+        arr = ak.arange(2**64 - 4, 2**64 - 1, dtype=int_types)
+        bv = ak.BitVector(arr, width=64)
+        assert isinstance(bv, client_dtypes.BitVector)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+    def test_bit_vector_error_handling(self):
+        # fail on argument types
+        with pytest.raises(TypeError):
+            ak.BitVector(17, width=4)
+        arr = ak.array([1.1, 8.3])
+        with pytest.raises(TypeError):
+            ak.BitVector(arr)
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_bit_vector_edge_case(self, int_types):
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 15 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 16 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 17 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 18 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+import pytest
+
+import arkouda as ak
+
+# from arkouda import client_dtypes
+
+INT_TYPES = [ak.int64, ak.uint64]
+
+
+class TestClientDType:
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:13]: Remove unused import and update class name to match function (i.e. make it plural `ClientDTypes`)

```suggestion

INT_TYPES = [ak.int64, ak.uint64]


class TestClientDTypes:
```

Comment (pierce314159) [reply PROTO_tests/tests/client_dtypes_test.py]: @jaketrookman you only did half of my suggestion. You didn't make the class name plural

Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated for real this time

File: PROTO_tests/tests/client_dtypes_test.py [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+        ips = ak.randint(1, 2**32, prob_size)
+        ip_list = ak.array(ips)
+        ipv4 = ak.IPv4(ip_list)
+        ip_as_dot = [ipaddress.IPv4Address(ip) for ip in ips.to_list()]
+        ip_as_int = [ipv4.normalize(ipd) for ipd in ip_as_dot]
+        assert ips.to_list() == ip_as_int
+
+    @pytest.mark.parametrize("size", pytest.prob_size)
+    def test_is_ipv4(self, size):
+        x = [random.getrandbits(32) for i in range(size)]
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:147]: half these tests parameterize `pytest.prob_size` as `prob_size` and the other half as `size`, we should pick one and try and be consistent.

Also I feel like `x` could get quite large for the client to handle depending on the value of `size`. I think it'd be safer to just pick a fixed size. Especially because we have several `.to_list` calls later in the test

Also `i` is never used, so we should replace it with `_` like what was suggested previous for `test_is_ipv6`

File: PROTO_tests/tests/client_dtypes_test.py [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+        with pytest.raises(TypeError):
+            ak.is_ipv4(ak.array(x, ak.float64))
+
+        with pytest.raises(RuntimeError):
+            ak.is_ipv4(ak.array(x, dtype=ak.uint64), ak.arange(2, dtype=ak.uint64))
+
+    @pytest.mark.parametrize("size", pytest.prob_size)
+    def test_is_ipv6(self, size):
+        x = [random.getrandbits(128) for _ in range(size)]
+        low = ak.array([i & (2**64 - 1) for i in x], dtype=ak.uint64)
```
Comment (pierce314159) [inline PROTO_tests/tests/client_dtypes_test.py:168]: same deal as with `test_is_ipv4`, with all the converting between client and server i vote we pick a fixed size for these arrays

**Review by pierce314159 — CHANGES_REQUESTED** [review]

File: PROTO_tests/tests/client_dtypes_test.py [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+    """
+    Note: BitVector operations are not tested here because the class is
+    only a wrapper on a pdarray to display as such.
+    The class does not actually store values as bit-values,
+    it only converts to a bit representation for display.
+    Thus, pdarray testing covers these operations.
+    """
+
+    @pytest.mark.parametrize("int_types", INT_TYPES)
+    def test_bit_vector_creation(self, int_types):
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+        # Test handling of python dotted-quad input
+        ipv4 = ak.ip_address(py_ips)
+        assert isinstance(ipv4, ak.IPv4)
+        assert ipv4.to_list() == py_ips
+        assert ipv4.dtype == ak.bitType
+
+    @pytest.mark.parametrize("prob_size", pytest.prob_size)
+    def test_ipv4_normalization(self, prob_size):
+        prob_size = 100
+        ips = ak.randint(1, 2**32, prob_size)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 26 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

File: PROTO_tests/tests/client_dtypes_test.py (hunk 27 above)
Comment (jaketrookman) [reply PROTO_tests/tests/client_dtypes_test.py]: Updated

**Review by pierce314159 — APPROVED** [review]
looks good!

**Comment by pierce314159** [comment]
I think once this gets the seal of approval from @Ethan-DeBandi99, we just need to merge the latest from master and it'll be good to go

**Review by Ethan-DeBandi99 — APPROVED** [review]
